rgw: Add rgw rate limiting per user and per bucket

ofriedma opened the pull request:
Add rgw rate limiting per user and per bucket

Signed-off-by: Or Friedmann <ofriedma@redhat.com>

dang commented on src/rgw/rgw_process.cc:
```diff
@@ -383,7 +391,17 @@ int process_request(rgw::sal::Store* const store,
+    userfind = "u" + userfind;
+    std::string bucketfind = !rgw::sal::Bucket::empty(s->bucket.get()) ? "b" + s->bucket->get_marker() : "";
+    const char *method = s->info.method;
+    std::cout << userfind << std::endl;
```
Remove debug output?

dang commented on src/rgw/rgw_rest.cc:
```diff
@@ -759,6 +760,20 @@ int dump_body(struct req_state* const s,
+  if(len > 0 && !healthchk) {
+    std::string userfind;
+    s->user->get_id().to_str(userfind);
+    userfind = "u" + userfind;
```
There's a lot of these open-coded string setups for getting keys.  Maybe a pair of helper functions would make this cleaner, and allow us to change if necessary in the future?

dang commented on src/rgw/rgw_sal.h:
```diff
@@ -217,6 +217,7 @@ class Store {
     virtual int register_to_service_map(const DoutPrefixProvider *dpp, const std::string& daemon_type,
 					const std::map<std::string, std::string>& meta) = 0;
     virtual void get_quota(RGWQuotaInfo& bucket_quota, RGWQuotaInfo& user_quota) = 0;
+    virtual void get_qos(RGWQoSInfo& bucket_qos, RGWQoSInfo& user_qos, RGWQoSInfo& anon_qos) = 0;
```
This needs to be implemented as a stub in DBStore.

cbodley reviewed (commented):
the choice of `tbb::concurrent_unordered_map` and atomics instead of locks imposes a lot of restrictions on what we can implement here - both in terms of garbage collection, and what we can track per entry. this design seems well suited for counting the number of ops in the last second, but not so much for tracking concurrent ops or bandwidth calculations

we talked about doing some kind of time-averaging for the bandwidth calculations. do you have any more thoughts about that part?

i have some nits about naming, formatting, etc. but i tried to focus on design issues in this first review

ofriedma commented:
@cbodley @dang @mattbenjamin 
I've pushed the new commit
Changes
"lazy" token bucket mechanism for bandwidth and ops estimations
GC for the data structure is based on size (Watermark) and not time
Changed it to unordered_map from tbb::concurrent_unordered_map + mutex for insertions/lookups
Using xattr for bucket/user limitation configuration

Because RGW will not drop running request even if it reached the bandwidth limit, I added a debt of maximum of 60 seconds, so the tenant will not be able to send more requests in maximum of 60 seconds (maybe we should make it configurable, what do you think?)

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,218 @@
+    int64_t read_ops = (tenant.max_read_ops * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t read_bw = (tenant.max_read_bytes * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t write_ops = (tenant.max_write_ops * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
+    int64_t write_bw = (tenant.max_write_bytes * ((double)(curr_timestamp - it->second.ts) / std::nano::den));
```
do we have a way to account for burst also? what would that look like?

my understanding is that we're refilling the buckets every second, meaning that we don't remember anything past 1 second. this really doesn't seem right to me, when it's common for requests to span several seconds or even minutes

i think it would be much more natural to think about these limits as rates per minute rather than second. then we could use a pure token bucket algorithm without relying on the CONCURR tracking piece. what do you think?

cbodley replied:
> Because RGW will not drop running request even if it reached the bandwidth limit, I added a debt of maximum of 60 seconds, so the tenant will not be able to send more requests in maximum of 60 seconds (maybe we should make it configurable, what do you think?)

would tracking the rates per-minute also address this concern?

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,241 @@
+      dout(0) << "time_in_ns " << time_in_ns << dendl;
+      it->second.ts = curr_timestamp;
+
+      it->second.write[RGW_QOS_ROUND_OPS] += static_cast<int64_t>(tenant.max_write_ops * time_in_ns * fixed_float) % fixed_float;
+      int round = it->second.write[RGW_QOS_ROUND_OPS] / fixed_float;
+      it->second.write[RGW_QOS_ROUND_OPS] = it->second.write[RGW_QOS_ROUND_OPS] - round * fixed_float;
```
instead of tracking `RGW_QOS_ROUND_OPS/BYTES` separately, what do you think about redefining the `RGW_QOS_OPS` and `RGW_QOS_BW` counters to be in multiples of 10000?

that would just mean multiplying `tenant.max_write_ops/bytes` by 10000 for comparisons, and dividing the counters by 10000 when you want their actual values:
```c++
int64_t qos_entry::read_ops() const { return read[RGW_QOS_OPS] / 10000; }
int64_t qos_entry::write_ops() const { return write[RGW_QOS_OPS] / 10000; }
```
this way, we only need the two counters (RGW_QOS_OPS/BW), and this math in `increase_tokens()` is a lot simpler

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,241 @@
+    auto ret = qos_entries.find(key);
+    if(ret == qos_entries.end())
+    {
+      insert_lock.lock();
+      ret = qos_entries.emplace(std::piecewise_construct,
+                                 std::forward_as_tuple(key),
+                                 std::forward_as_tuple()).first;
+      insert_lock.unlock();
```
i don't see how it could be safe to only hold the lock over `emplace()`, but not over `find()` and `size()`

ofriedma commented:
@cbodley I think I've covered everything, would you mine to review again?

Thank you

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+// once the active has reached the watermark for clearing it will call the replace_active() thread using cv
+// The replace_active will clear the previous QosDatastruct after all requests to it has been done (use_count() > 1)
+// In the meanwhile new requests will come into the newer active
+class QosActiveDatastruct {
```
"ActiveRateLimiter" maybe?

cbodley commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,72 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
 }
+bool rate_limit(rgw::sal::Store* store, req_state* s) {
+  // we dont want to limit health check or system or admin requests
+  auto is_admin_or_system = s->user->get_info();
```
`const auto&` to avoid making a copy of RGWUserInfo

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+      return increase_entry_write(it, qos_info.max_write_ops, qos_info.max_write_bytes);
+    }
+    bool increase_bw(const char *method, std::string tenant, int64_t amount, const RGWQoSInfo& info) {
+      if(tenant.empty() || tenant.length() == 1 || !info.enabled)
```
rgw already has a very specific meaning for "tenant", so we should find a different term in the context of rate limiting. here, i would probably just call this the "key" since it's what we use to access the map

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+    QosActiveDatastruct(QosActiveDatastruct&&) = delete;
+    QosActiveDatastruct& operator =(QosActiveDatastruct&&) = delete;
+    QosActiveDatastruct() {
+      qos[0] = std::shared_ptr<QosDatastruct>(new QosDatastruct(&replacing,&cv));
+      qos[1] = std::shared_ptr<QosDatastruct>(new QosDatastruct(&replacing,&cv));
```
`std::make_shared<QosDatastruct>(&replacing, &cv);`

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+#include <condition_variable>
+#include "rgw_common.h"
+
+#if !defined(dout_subsys)
+#define dout_subsys ceph_subsys_rgw
+#define def_dout_subsys
+#endif
+
+#define dout_context g_ceph_context
```
it's best not to mess with dout macros in header files. i don't think QosDatastruct needs to do any logging. for QosActiveDatastruct, you can either move its implementation into a rgw_qos.cc file and use the macros there, or add a `DoutPrefixProvider*` argument to its constructor and do the logging with `ldpp_dout()` which doesn't depend on the `dout_subsys`/`dout_context` macros

ofriedma replied:
I think we would want to get logging from QosDatastruct(Ratelimiter), maybe with even higher logging level (30) so in case of bug in the rate limiting the user will be able to bring us logs that could help us debug it.
I will add to both classes the DoutPrefixProvider*

cbodley replied:
ok. i think `QosActiveDatastruct` should be its own provider since it's running a background thread, and all of the log output should look the same. to do that,
```c++
class QosActiveDatastruct : public DoutPrefix {
 public:
  QosActiveDatastruct(CephContext* cct)
    : DoutPrefix(cct, ceph_subsys_rgw, "rate limiter: ") {
```
for logging inside of `QosDatastruct`, we'll be calling this from the context of an `RGWOp`, so want to use that op's prefix. that means the caller will need to pass in the `const DoutPrefixProvider* dpp` as an argument to increase_entry()/increase_bw() (and not to its constructor)

log statements within `QosActiveDatastruct` would use `ldpp_dout(this, 20) <<`, and `ldpp_dout(dpp, 20) <<` in `QosDatastruct`

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+    QosDatastruct& operator =(const QosDatastruct&) = delete;
+    QosDatastruct(QosDatastruct&&) = delete;
+    QosDatastruct& operator =(QosDatastruct&&) = delete;
+    QosDatastruct() {};
```
should not have a default constructor. this would leave `replacing` and `cv` uninitialized

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+    QosDatastruct(QosDatastruct&&) = delete;
+    QosDatastruct& operator =(QosDatastruct&&) = delete;
+    QosDatastruct() {};
+    QosDatastruct(std::atomic_bool *replacing, std::condition_variable *cv) {
+      // prevent rehash, so no iterators invalidation
+      qos_entries.max_load_factor(1000);
+      this->replacing = replacing;
+      this->cv = cv;
```
constructors should use the [member initializer list](https://en.cppreference.com/w/cpp/language/constructor) to initialize its variables:
```c++
    QosDatastruct(std::atomic_bool *replacing, std::condition_variable *cv)
      : replacing(replacing), cv(cv) {
```

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+  std::atomic_bool *replacing;
+  typedef std::unordered_map<std::string, qos_entry> hash_map;
+  hash_map qos_entries{map_size};
+  bool is_read_op(const char *method) {
```
can be a static function, since it doesn't depend on any member variables. i also prefer std::string_view for comparisons:
```c++
static bool is_read_op(std::string_view method) {
  return method == "GET" || method == "HEAD";
}
```

cbodley commented on src/rgw/rgw_rest.cc:
```diff
@@ -781,7 +792,18 @@ int recv_body(struct req_state* const s,
 {
   try {
-    return RESTFUL_IO(s)->recv_body(buf, max);
+    int len = RESTFUL_IO(s)->recv_body(buf, max);
```
only the call to `recv_body()` should be covered by the try-catch. that means moving `int len` out of the try block

cbodley reviewed (commented):
cool, the logic is looking much better! focusing review more on design, naming, style, etc

ofriedma commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
+    if(curr_timestamp > it->second.ts && !it->second.first_run && minimum_time_reached(curr_timestamp, it))
+    {
+      int64_t time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(curr_timestamp - it->second.ts).count() / 60.0 / std::milli::den * fixed_float; // / 60 to make it work with 1 min token bucket
+      it->second.ts = curr_timestamp;
+      int64_t write_ops = tenant.max_write_ops * time_in_ms;
+      int64_t write_bw = tenant.max_write_bytes * time_in_ms;
+      int64_t read_ops = tenant.max_read_ops * time_in_ms;
+      int64_t read_bw = tenant.max_read_bytes * time_in_ms;
```
Yes, I have tested and it is working fine with 1(probably because it's combined with higher minimum_time_reached).
Before this change for lower values after 61 seconds of running there were not enough tokens and the tenant still got only 1 request (it took 70~ seconds to get 2 requests)

ofriedma commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+
+    // find or create an entry, and return its iterator
+  auto find_or_create(const std::string& key) {
+    std::shared_lock rlock(insert_lock);
```
I see a small improvement using rwlock here (when running with single tenant which is the best case for rwlock), we can KIS and use a non shared mutex anyway.
What do you think?

cbodley replied:
cool, thanks for checking. let's leave it as is 👍

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+                       const RGWQoSInfo& tenant)
+  {
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
+    if(curr_timestamp > it->second.ts && !it->second.first_run && minimum_time_reached(curr_timestamp, it))
```
(this is important because you're accessing `it->second.ts` here which is uninitialized until that first_run check)

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,246 @@
+      this->replacing = replacing;
+      this->cv = cv;
+    };
+    bool increase_entry(const char *method, std::string tenant, ceph::coarse_real_time curr_timestamp, const RGWQoSInfo& qos_info) {
```
or maybe `check()` or `check_op()` that returns false if the request should be denied?

ofriedma commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,72 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+      user_info = global_anon;
+  }
+  bool limit_bucket = false;
+  bool limit_user = s->qos_data->increase_entry(method, s->ratelimit_user_name, s->time, user_info);
+  if(!rgw::sal::Bucket::empty(s->bucket.get()))
+    limit_bucket = s->qos_data->increase_entry(method, s->ratelimit_bucket_name, s->time, bucket_info);
```
"if the user limit check fails, we shouldn't call into the bucket check and modify its counters too" - correct, I will change it.
I'm not sure I understood, why the user bucket order matters?

cbodley commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,79 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+      *user_qos = global_anon;
+  }
+  bool limit_bucket = false;
+  bool limit_user = s->qos_data->should_rate_limit(method, s->ratelimit_user_name, s->time, user_qos, s);
+  if(!rgw::sal::Bucket::empty(s->bucket.get()) && !limit_user)
+    limit_bucket = s->qos_data->should_rate_limit(method, s->ratelimit_bucket_name, s->time, bucket_qos, s);
```
this would be a good place to log why a request is being rejected. for example, something like
> bucket rate limit exceeded (ops=X bw=Y)

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,238 @@
+    ceph::timespan ts;
+    bool first_run = true;
+    std::mutex ts_lock;
+    const int64_t read_ops()
```
`int64_t read_ops() const {` to signal that the functions don't modify anything

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,218 @@
+    RGW_QOS_CONCURR,
+    RGW_QOS_BW,
+  };
+  struct qos_entry {
```
would still like to see some unit testing of the `qos_entry` logic (this is why the pr is tagged `needs-test`)

the simulator helps us measure performance, but it won't run as part of 'make check', so won't help us catch any regressions

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,238 @@
+
+
+class RateLimiter {
+  static constexpr uint64_t fixed_float_rgw_ratelimit = 1000;
+  struct counters {
+    int64_t ops = 0;
+    int64_t bw = 0;
```
please comment that the counters are tracked in multiples of `fixed_float_rgw_ratelimit`

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,238 @@
+     This function is necessary to force the increase tokens add at least 1 token when it updates the last stored timestamp.
+     That way the user/bucket will not lose tokens because of rounding
+  */
+  const bool minimum_time_reached(const ceph::timespan curr_timestamp, const qos_entry& it)
```
please make the function itself const, not the return value

cbodley commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,238 @@
+      qos_entries.max_load_factor(1000);
+    };
+
+    bool should_rate_limit(const char *method, const std::string& key, ceph::coarse_real_time curr_timestamp, const RGWQoSInfo* qos_info, const DoutPrefixProvider* dpp) {
```
function takes `DoutPrefixProvider` but doesn't use it

cbodley reviewed (commented):
the commit includes an unintended change to the zstd submodule

yuvalif commented on src/rgw/rgw_admin.cc:
```diff
@@ -1260,6 +1291,50 @@ static bool dump_string(const char *field_name, bufferlist& bl, Formatter *f)
+    case OPT::RATELIMIT_SET:
+    case OPT::GLOBAL_RATELIMIT_SET:
+      if (have_max_read_ops) {
+        if (max_read_ops > -1) {
```
is -1 a special value that the use can pass to the commandline, or is this a default value?

yuvalif commented on src/rgw/rgw_common.h:
```diff
@@ -690,6 +694,43 @@ void decode_json_obj(rgw_placement_rule& v, JSONObj *obj);
+
+class RateLimiter;
+struct RGWQoSInfo {
+  int64_t max_write_ops;
```
nit:
```suggestion
  int64_t max_write_ops = 0;
```
and remove ctor

yuvalif commented on src/rgw/rgw_common.h:
```diff
@@ -690,6 +694,43 @@ void decode_json_obj(rgw_placement_rule& v, JSONObj *obj);
 }
+
+class RateLimiter;
+struct RGWQoSInfo {
```
maybe we should move this into its own file?
why do we add more stuff to rgw_common?

ofriedma replied:
https://github.com/ceph/ceph/pull/42891#discussion_r694915814

yuvalif commented on src/rgw/rgw_main.cc:
```diff
@@ -534,6 +534,9 @@ int radosgw_Main(int argc, const char **argv)
 
   OpsLogManifold *olog = new OpsLogManifold();
+  ActiveRateLimiter ratelimiting{cct.get()};
+  ratelimiting.start();
```
do we need a `ratelimiting.stop();` ?

ofriedma replied:
@cbodley what do you think?

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+bool rate_limit(rgw::sal::Store* store, req_state* s) {
+  // we dont want to limit health check or system or admin requests
+  const auto& is_admin_or_system = s->user->get_info();
+  if (s->op_type ==  RGW_OP_GET_HEALTH_CHECK || is_admin_or_system.admin || is_admin_or_system.system)
```
style: please use brackets

ofriedma replied:
ok

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+  RGWQoSInfo global_anon;
+  RGWQoSInfo* bucket_qos;
+  RGWQoSInfo* user_qos;
+  store->get_qos(global_bucket, global_user, global_anon);
```
what if all qos info object has a "disabled" flag, why not bail out here?

ofriedma replied:
because there are the global values and the per bucket and user values which should be take in place anyway (the global will take in place only if the specific values are disabled).

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+  RGWQoSInfo global_user;
+  RGWQoSInfo global_bucket;
+  RGWQoSInfo global_anon;
+  RGWQoSInfo* bucket_qos;
```
you can use a reference instead of a pointer here, and initialize it after you fetched the global ones

yuvalif replied:
or even better, set the values directly to: `s->user_ratelimit`

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+  const auto& is_admin_or_system = s->user->get_info();
+  if (s->op_type ==  RGW_OP_GET_HEALTH_CHECK || is_admin_or_system.admin || is_admin_or_system.system)
+    return false;
+  std::string userfind;
```
why do you need this temporary variable?
can you just set the value to `s->ratelimit_user_name`?

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+  s->user->get_id().to_str(userfind);
+  userfind = "u" + userfind;
+  s->ratelimit_user_name = userfind;
+  std::string bucketfind = !rgw::sal::Bucket::empty(s->bucket.get()) ? "b" + s->bucket->get_marker() : "";
```
same as: https://github.com/ceph/ceph/pull/42891/files#r767575747

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+      if (user_qos_temp.enabled)
+        *user_qos = user_qos_temp;
+    } catch (buffer::error& err) {
+      dout(0) << "ERROR: failed to decode rate limit" << dendl;
```
any reason why not using `ldpp_dout`?

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -87,12 +90,83 @@ void RGWProcess::RGWWQ::_process(RGWRequest *req, ThreadPool::TPHandle &) {
+  }
+  bool limit_bucket = false;
+  bool limit_user = s->qos_data->should_rate_limit(method, s->ratelimit_user_name, s->time, user_qos);
+  if(!rgw::sal::Bucket::empty(s->bucket.get()) && !limit_user) {
```
you already checked if there is a bucket in line 129

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -169,6 +243,11 @@ int rgw_process_authenticated(RGWHandler_REST * const handler,
   op->pre_exec();
 
+  ldpp_dout(op, 2) << "check rate limiting" << dendl;
+  bool islimited = rate_limit(store, s);
```
nit:
```suggestion
  if (rate_limit(store, s)) {
    return -ERR_RATE_LIMITED;
  }
```

yuvalif commented on src/rgw/rgw_process.h:
```diff
@@ -38,6 +38,8 @@ struct RGWProcessEnv {
   std::string uri_prefix;
   std::shared_ptr<rgw::auth::StrategyRegistry> auth_registry;
+  //maybe there is a better place to store the rate limit data structure
+  ActiveRateLimiter* ratelimiting;
```
if the owner is a local variable in rgw_main then this is fine to leave it as a bare pointer here

yuvalif replied:
if the owner is a local variable in rgw_main then this is fine to leave it as a bare pointer here

yuvalif commented on src/rgw/rgw_process.cc:
```diff
@@ -194,6 +273,7 @@ int process_request(rgw::sal::Store* const store,
 		    rgw::dmclock::Scheduler *scheduler,
                     string* user,
                     ceph::coarse_real_clock::duration* latency,
+                    std::shared_ptr<RateLimiter> ratelimit,
```
why do you need a `shared_ptr`?
see: https://github.com/ceph/ceph/pull/42891/files#r767585678

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+
+
+class RateLimiter {
+  static constexpr int64_t fixed_float_rgw_ratelimit = 1000;
```
please specify the unit here

ofriedma replied:
There is no unit in here, it is basically how many digits right of the decimal point (3 digits for 1000)
It is increasing our precision when the bytes or ops limit are too low and the tokens added are 0 because of that.
this is why read_ops() for example is the total ops left / 1000 integral value

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+  // counters are tracked in multiples of fixed_float_rgw_ratelimit
+  struct counters {
+    int64_t ops = 0;
+    int64_t bw = 0;
```
this is "bytes" not "bandwidth" (=bytes/sec)

ofriedma replied:
ok

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+  };
+  static constexpr size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::atomic_bool *replacing;
```
please use a reference and not a pointer

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+  static constexpr size_t map_size = 2000000; // will create it with the closest upper prime number
+  std::shared_mutex insert_lock;
+  std::atomic_bool *replacing;
+  std::condition_variable *cv;
```
please use a reference

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+  }
+
+  bool should_rate_limit_read(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
```
do we use "0" or "-1" as unlimited?
if we use zero, we probably need to add the to the help in radosgw-admin
also, not sure we have "-1" there? or why do we use signed long long?

ofriedma replied:
0 means unlimited I will add it to the radosgw-admin -h
so ops_limit could not  be negative and bw_limit(info->max_read_bytes) could not be too.
info->max_read_bytes will force us to cast it to int64_t each decrease_bytes() (formerly increase_bw()) call as we do **not** want implicit casting to uint64
we can change ops_limit to be uint64, but I'm not sure if we want to seperate their types.
@cbodley what do you think?

cbodley replied:
right, since `int64_t bytes` can go negative due to burst, i think it makes sense for the ops and limits to be expressed in `int64_t` as well. that avoids signed vs. unsigned comparisons, and the logic can be (mostly) identical for ops and bw calculations

i do have a question related to `0 = unlimited`, though.. if `RGWQoSInfo` defaults everything to 0/unlimited, then what's the point of the extra field `RGWQoSInfo::enabled`?

ofriedma replied:
Honestly it took me a minute or two to remember why 😌.
There this enable or disable to handle the global and not global, for example I want that a bucket will have specifically no limit and another one will be as the global settings, I will enable on the bucket the policy and then the rgw will take it and not the global and on the other bucket it will take the global

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+
+  bool should_rate_limit_read(qos_entry& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it.read_ops() - 1 < 0) && (ops_limit > 0)) ||
```
why not: `it.read_ops() == 0` ? can we have negative values there?

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      cv.wait(lk);
+      current_active = current_active ^ 1;
+      ldpp_dout(this, 20) << "replacing active qos data structure" << dendl;
+      while (!stopped && qos[(current_active ^ 1)].use_count() > 1 ) {
```
what if `use_count()` never goes to 1? as requests keep on arriving

ofriedma replied:
Because of this line:       current_active = current_active ^ 1;
New requests will come to the other map as it is swapping the active map

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+    bool first_run = true;
+    std::mutex ts_lock;
+    // Those functions are returning the integer value of the tokens 
+    int64_t read_ops () const
```
do we return ops/second or ops?

ofriedma replied:
ops (tokens)

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+  void increase_tokens(qos_entry& it, const ceph::timespan curr_timestamp,
+                       const RGWQoSInfo* info)
+  {
+    constexpr int fixed_float = fixed_float_rgw_ratelimit;
```
why "float"?

ofriedma replied:
correct - the right name should be fixed_point

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      return false;
+    return true;
+  }
+  void increase_tokens(qos_entry& it, const ceph::timespan curr_timestamp,
```
```suggestion
  void increase_tokens(qos_entry& it, const ceph::timespan& curr_timestamp,
```

ofriedma replied:
As mentioned in the above review:
"ceph::timespan is just a uint64_t, pass it by value"

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      // it is possible that some requests will have the same ts, because we are using coarse real time
+      auto& it = find_or_create(key);
+      auto curr_ts = curr_timestamp.time_since_epoch();
+      std::unique_lock lock(it.ts_lock);
```
maybe better to move the record lock inside the functions, and use it only when entry values are modified?

ofriedma replied:
we will need RWlocks anyway, I don't see how it helps, values modification here is quite common

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      if(key.empty() || key.length() == 1 || !info->enabled)
+        return;
+      bool is_read = is_read_op(method);
+      if ((is_read && !info->max_read_bytes) || (!is_read && !info->max_write_bytes))
```
```suggestion
      if ((is_read && info->max_read_bytes == 0) || (!is_read && info->max_write_bytes == 0))
```
would be even better if you used a constant instead of zero here (and in all other places)

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      }
+      
+    }
+    void increase_bw(const char *method,const std::string& key, int64_t amount, const RGWQoSInfo* info, const DoutPrefixProvider* dpp) {
```
shouldn't this function be called "decrease_bw"? (or "decrease_bytes")

ofriedma replied:
correct - will rename it

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      std::unique_lock lock(it.ts_lock);
+      // we don't want the tenant to be with higher debt than 120 seconds(2 min) of its limit
+      if (is_read)
+        it.read.bw = std::max(it.read.bw - amount * fixed_float_rgw_ratelimit,info->max_read_bytes * fixed_float_rgw_ratelimit * -2);
```
maybe define `info->max_read_bytes * fixed_float_rgw_ratelimit * -2` as "burst size"?

ofriedma replied:
info->max_read_bytes can be changed inflight and it appears only once in the code, I'm not sure we should define it

ofriedma commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+    }
+    ~ActiveRateLimiter() {
+      ldpp_dout(this, 20) << "stopping ratelimit_gc thread" << dendl;
+      cv_m.lock();
```
Without locking this variable there are situations that the notify will arrive before the wait and the thread will stuck on wait()

cbodley replied:
to avoid those races between notify and wait, the mutex should be locked over the call to `notify_all()`. `stopped = true` doesn't necessarily need to happen inside the lock, but it might as well be

ofriedma commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+        if (stopped && cv.wait_for(lk, 1min) != std::cv_status::timeout)
+          return;
+      }
+      if (stopped)
```
if the stopped flag has became true we will skip the clear()

ofriedma commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+      qos[0] = std::make_shared<RateLimiter>(&replacing, &cv);
+      qos[1] = std::make_shared<RateLimiter>(&replacing, &cv);
+    }
+    ~ActiveRateLimiter() {
```
@cbodley what do you think?

cbodley replied:
@yuvalif makes a good point, order of initialization/shutdown can be messy and lead to crashes. we've seen a lot of those issues around `RGWRados::initialize()/finalize()` in the past

but on the other hand, i think [RAII](https://en.cppreference.com/w/cpp/language/raii) is the best tool that C++ provides for managing lifetimes, and triggering this shutdown from `~ActiveRateLimiter()` is the RAII way to do it

so the question here is whether `ActiveRateLimiter` carries any external dependencies that could be destroyed before it is. it looks like `CephContext` is the only dependency, and the stack of `radosgw_Main()` manages the lifetime of `CephContext` with RAII also (`auto cct = global_init(...)`, which returns `boost::intrusive_ptr<CephContext>`)

so i think @ofriedma's approach is the most natural way to handle init/shutdown order, and wouldn't recommend manual shutdown unless we actually see issues here

cbodley commented:
one of the make check failures is from `/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/cli/radosgw-admin/help.t: failed`. you can see the diff output if you click through the `Details`

cbodley commented:
another failure is coming from the decode of RGWPeriod:
```
        /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/RGWPeriod
skipping unrecognized type RGWQuotaInfo
error: End of buffer
error: End of buffer
**** failed to decode /home/jenkins-build/build/workspace/ceph-pull-requests/ceph-object-corpus/archive/15.0.0-539-g191ab33faf/objects/RGWPeriod/6b09071de3802c158d8624cac6fe3898 ****
```
i'm not sure what's missing here for `RGWQuotaInfo`

cbodley commented on src/test/rgw/test_ratelimit_unittest.cc:
```diff
@@ -0,0 +1,191 @@
+  auto time = ceph::coarse_real_clock::now();
+  std::string key = "uuser123";
+  bool success = active->should_rate_limit("GET", key, time, &info);
+  sleep(61);
+  time = ceph::coarse_real_clock::now();
```
you can use fake times instead of `sleep()`+`now()`! for example,
```c++
const auto t0 = ceph::coarse_real_clock::now();
...
const auto t1 = t0 + 60s; // check again after 60 seconds
```

cbodley commented on src/test/rgw/CMakeLists.txt:
```diff
@@ -52,6 +52,15 @@ set(test_rgw_a_src test_rgw_common.cc)
+add_executable(test_rgw_ratelimit_gc test_ratelimit_gc.cc)
+target_link_libraries(test_rgw_ratelimit_gc ${rgw_libs} Boost::coroutine)
+
+add_executable(unittest_rgw_ratelimit test_ratelimit_unittest.cc)
```
nit: i would call this one `test_ratelimit`, and the others `bench_ratelimit{_gc}`

cbodley commented on src/test/rgw/test_ratelimit_unittest.cc:
```diff
@@ -0,0 +1,191 @@
+TEST(RGWRateLimit, shouldratelimit1)
+{
+  // info.enabled = false, so no limit
+  std::unique_ptr<CephContext> cct = std::make_unique<CephContext>(CEPH_ENTITY_TYPE_ANY);
```
tests shouldn't need to create their own `CephContext`. you can just add ` $<TARGET_OBJECTS:unit-main>` to the executable for access to `g_ceph_context`. same for test_ratelimit.cc and test_ratelimit_gc.cc

you can find examples of this in src/test/rgw/CMakeLists.txt, like `unittest_rgw_compression`. the `unit-main` target just adds `src/test/unit.cc`, which has a `main()` function that calls `global_init()` and `common_init()` to initialize `g_ceph_context`

ofriedma replied:
It looks like the main function on the bench_* files is duplicated with the one from src/test/unit.cc

cbodley commented on src/test/rgw/test_ratelimit_unittest.cc:
```diff
@@ -0,0 +1,191 @@
+{
+  // info.enabled = false, so no limit
+  std::unique_ptr<CephContext> cct = std::make_unique<CephContext>(CEPH_ENTITY_TYPE_ANY);
+  std::shared_ptr<ActiveRateLimiter> qos(new ActiveRateLimiter(cct.get()));
+  qos->start();
+  auto active = qos->get_active();
```
since this is unit testing `qos_entry` specifically, we shouldn't need to deal with `ActiveRateLimiter` or `RateLimiter` at all. that's what i was getting at with my earlier [comment](https://github.com/ceph/ceph/pull/42891/#discussion_r719708380):

> i'd really like to see this factored out into its own class so we can write unit tests against it

that would mean splitting the counter logic out of `RateLimiter` and into functions on `qos_entry` (either as member functions, or free functions taking `qos_entry&`). so ideally, we'd have `class ActiveRateLimiter` that only deals with the GC, `class RateLimiter` that only deals with the map of entries, and `struct qos_entry` that only deals with the counters

that way these unit tests can ignore `ActiveRateLimiter`/`RateLimiter` entirely, unless we decide to add test cases that exercise those too

cbodley commented on src/test/rgw/test_ratelimit_unittest.cc:
```diff
@@ -0,0 +1,191 @@
+
+TEST(RGWRateLimit, shouldratelimit1)
+{
+  // info.enabled = false, so no limit
```
these test comments are helpful, but i think they'd be more helpful as the test names. for example,
```
TEST(RGWRateLimit, op_limit_not_enabled)
TEST(RGWRateLimit, reject_op_over_limit)
TEST(RGWRateLimit, accept_op_after_refill)
TEST(RGWRateLimit, reject_bw_over_limit)
```
etc

cbodley reviewed (commented):
nice work on the unit test coverage 👍

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+
+
+class RateLimiter {
+  static constexpr int64_t fixed_point_rgw_ratelimit = 1000;
```
please add an example of the calculation in a comment

yuvalif commented on src/rgw/rgw_qos.h:
```diff
@@ -0,0 +1,256 @@
+    }
+    else if(curr_timestamp > it.ts && minimum_time_reached(curr_timestamp, it))
+    {
+      int64_t time_in_ms = std::chrono::duration_cast<std::chrono::milliseconds>(curr_timestamp - it.ts).count() / 60.0 / std::milli::den * fixed_point; // / 60 to make it work with 1 min token bucket
```
please add const to local vars

cbodley commented on src/rgw/rgw_zone.h:
```diff
@@ -1028,18 +1028,28 @@ struct RGWPeriodConfig
     decode(user_quota, bl);
+    decode(bucket_qos, bl);
+    decode(user_qos, bl);
+    decode(anon_qos, bl);
```
tthe ceph-object-corpus test is failing on `RGWPeriod` because the encode/decode of `RGWPeriodConfig` changed without bumping the version number

ofriedma commented:
Added docs
Added tests to RateLimiter ActiveRateLimiter and to RateLimiterEntry
Removed from the PR the qos keyword
Moved some functions to RateLimiterEntry and kept only interfaces of them inside RateLimiter.

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+=====================
+
+The Ceph Object Gateway enables you to set rate limits on users and buckets. 
+rate limit include the maximum number of read ops and write ops per minute
```
```suggestion
rate limit includes the maximum number of read ops and write ops per minute
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+
+The Ceph Object Gateway enables you to set rate limits on users and buckets. 
+rate limit include the maximum number of read ops and write ops per minute
+and how many write bytes or read bytes per minute a user or bucket can make.
```
```suggestion
and how many bytes per minute could be written or read per user or bucket.
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+The Ceph Object Gateway enables you to set rate limits on users and buckets. 
+rate limit include the maximum number of read ops and write ops per minute
+and how many write bytes or read bytes per minute a user or bucket can make.
+Requests that are using GET or HEAD method in the REST request considered as read, otherwise it is write.
```
```suggestion
Requests that are using GET or HEAD method in the REST request are considered as "read requests", otherwise they are considered as "write requests".
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+rate limit include the maximum number of read ops and write ops per minute
+and how many write bytes or read bytes per minute a user or bucket can make.
+Requests that are using GET or HEAD method in the REST request considered as read, otherwise it is write.
+Every Object Gateway will track after the user/bucket usage itself and will not share those metrics with other gateways.
```
```suggestion
Every Object Gateway tracks per user/bucket separately, and these metrics are not shared metrics other gateways.
Also, note that usage is calculated after an operation was successfully finished.
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+Requests that are using GET or HEAD method in the REST request considered as read, otherwise it is write.
+Every Object Gateway will track after the user/bucket usage itself and will not share those metrics with other gateways.
+That means that the desired limits configured should be divide by the number of active Object Gateways.
+if userA should be with a limit 10 ops per minute and there are 2 Object Gateways in the cluster,
```
```suggestion
For example, if userA should be limited by 10 ops per minute and there are 2 Object Gateways in the cluster,
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+ 
+- **Rate Limit Scope:** The ``--ratelimit-scope `` option sets the scope for the rate limit.
+  The options are ``bucket`` , ``user`` and ``anonymous``. Bucket rate limit apply to buckets. 
+  User quotas apply to a user. Anonymous apply to an unauthenticated user.
```
```suggestion
  The user rate limit applies to a user. Anonymous apply to an unauthenticated user.
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+That means that the desired limits configured should be divide by the number of active Object Gateways.
+if userA should be with a limit 10 ops per minute and there are 2 Object Gateways in the cluster,
+the limit over userA should be 5 (10 ops per minute / 2 RGWs).
+
```
please add a note here, explaining that if the requests are not balanced between RGWs, the rate limit may be underutilized.

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+- **Bucket:** The ``--bucket`` option allows you to specify a rate limit for
+  bucket.
+
+- **Maximum Read Ops:** The ``--max-read-ops`` setting allows you to specify
```
is it mandatory to set at least one of the "max" values?

ofriedma replied:
It is mandatory, changed the cli examples to reflect it

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+the limit over userA should be 5 (10 ops per minute / 2 RGWs).
+
+
+- **Bucket:** The ``--bucket`` option allows you to specify a rate limit for
```
is this mandatory when the scope is "bucket"?

ofriedma replied:
yes , but I don't think it is worth mentioning because it's on the bucket rate limit set

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+  The options are ``bucket`` , ``user`` and ``anonymous``. Bucket rate limit apply to buckets. 
+  User quotas apply to a user. Anonymous apply to an unauthenticated user.
+  Anonymous scope is only available for global rate limit.
+
```
what about `--uid` ?

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+Enable/Disable User Rate Limit
+------------------------------
+
+Once you set a user rate limit, you may enable it. For example:: 
```
what if you set the global rate limits and then enable the user? will the global settings apply here?

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+Set Bucket Rate Limit
+---------------------
+
+Before you enable a rate limit, you must first set the rate limit parameters.
```
same as https://github.com/ceph/ceph/pull/42891/files#r771946110

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+Reading / Writing Global Rate Limit Configuration
+-------------------------------------------------
+
+You can read and write global rate limit settings in the period configuration. To
```
what is the relationship between the global and the per user/bucket limits?
do we block when any of them is reached, or do we override the values with the more specific ones?

mattbenjamin replied:
This doesn't seem to be resolved yet... I don't see an explanation of how global and per user/bucket limits interact, and I agree it's important to explain this.

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+A 0 value for num ops and / or num bytes means that the
+specific rate limit attribute check is disabled.
+
+Get Bucket Rate Limit
```
please specify the relationship between the use/bukcet ones?
do we block when the first one is reached?

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
+	radosgw-admin global ratelimit set --ratelimit-scope bucket --max-read-ops=1024
+	radosgw-admin global ratelimit enable --ratelimit-scope bucket
+
+The global rate limit can configure rate limit scope for all unauthenticated users::
```
can we also configure them for all users if we use:
```
radosgw-admin global ratelimit enable --ratelimit-scope=user
```

ofriedma replied:
yes

yuvalif replied:
so, can you please add this information/example?

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,157 @@ commands. ::
    be restarted for the changes to take effect.
 
 
+Rate Limit Management
```
are these configuration per tenant or across tenants?

ofriedma replied:
per tenant

yuvalif replied:
so please add this information in the doc

ofriedma commented:
jenkins render docs

ofriedma commented:
jenkins render docs

mattbenjamin commented on src/rgw/rgw_admin.cc:
```diff
@@ -1261,6 +1292,57 @@ static bool dump_string(const char *field_name, bufferlist& bl, Formatter *f)
+    case OPT::GLOBAL_RATELIMIT_ENABLE:
+      ratelimit.enabled = true;
+      break;
+
```
what is the ws for?

mattbenjamin commented on src/rgw/rgw_sal.h:
```diff
@@ -303,7 +303,7 @@ class Store {
     /** Get default quota info.  Used as fallback if a user or bucket has no quota set*/
     virtual void get_quota(RGWQuotaInfo& bucket_quota, RGWQuotaInfo& user_quota) = 0;
-    /** Enable or disable a set of bucket.  e.g. if a User is suspended */
+    virtual void get_ratelimit(RGWRateLimitInfo& bucket_ratelimit, RGWRateLimitInfo& user_ratelimit, RGWRateLimitInfo& anon_ratelimit) = 0;
```
the comment on  set_buckets_enabled is being lost

mattbenjamin reviewed (commented):
lgtm

ofriedma commented:
@cbodley 
Changed the docs as requested aded more details about global and regular configuration.
Will push in a separated PR the Admin API too.

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,171 @@ commands. ::
+if the requests are ``not`` balanced between RGWs, the rate limit may be underutilized.
+For example, if the ops limit is 5 and there are 2 RGWs, ``but`` the Load Balancer send load only to one of those RGWs,
+The effective limit would be 5 ops, because this limit is enforced per RGW.
+if there is a limit reached for bucket not for user or vice versa the request would be cancelled as well.
```
```suggestion
The request would be rejected if either the user limits or the bucket limits are reached.
```

yuvalif commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,171 @@ commands. ::
+For example, if the ops limit is 5 and there are 2 RGWs, ``but`` the Load Balancer send load only to one of those RGWs,
+The effective limit would be 5 ops, because this limit is enforced per RGW.
+if there is a limit reached for bucket not for user or vice versa the request would be cancelled as well.
+It means, if userA has 1 byte read limit per minute and this user tries to GET 1 GB object, the user will be able to do it.
```
This does not seem related to the previous sentence. Isn't that the definition of the maximum burst you allow?

ofriedma commented:
jenkins test make check

ofriedma commented:
Added admin API for the rate limiting

cbodley commented:
jenkins render docs

cbodley reviewed (approved):
i still want to see the rendered docs, but everything else looks wonderful!

cbodley commented on src/pybind/mgr/CMakeLists.txt:
```diff
@@ -13,7 +13,7 @@ if(WITH_MGR_ROOK_CLIENT)
 if(WITH_TESTS)
   include(AddCephTest)
-  add_tox_test(mgr ${CMAKE_CURRENT_SOURCE_DIR} TOX_ENVS py3 py37 mypy flake8 jinjalint nooptional)
+  add_tox_test(mgr ${CMAKE_CURRENT_SOURCE_DIR} TOX_ENVS py3 python3.7 mypy flake8 jinjalint nooptional)
```
i don't understand what these changes are doing. can we find someone familiar with python and ceph-mgr to review this part? maybe @epuertat @leseb?

ofriedma replied:
jenkins test make check

ofriedma replied:
Without this commit there is an error in the make check:
  py3: commands succeeded
SKIPPED:  py37: InterpreterNotFound: python3.7
ERROR:   mypy: commands failed

Anyway it's not mandatory to add this commit to the PR so once the make check is fine I'll drop it

ofriedma commented:
jenkins test make check

ofriedma commented:
jenkins test docs

ofriedma commented:
jenkins render docs

ofriedma commented:
jenkins render docs

cbodley commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,174 @@ commands. ::
+- **Maximum Write Bytes:** The ``--max-write-bytes`` setting allows you to specify
+  the maximum number of write bytes per minute per RGW. A 0 value disables this setting (which means unlimited access).
+ 
+- **Rate Limit Scope:** The ``--ratelimit-scope `` option sets the scope for the rate limit.
+  The options are ``bucket`` , ``user`` and ``anonymous``. Bucket rate limit apply to buckets. 
```
the extra space in `--ratelimit-scope ` is causing a rendering issue

ofriedma replied:
I've removed the extra space

cbodley commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,174 @@ commands. ::
+Get User Rate Limit
+-------------------
+
+Get the current configured rate limit parameters
+For example:: 
+
+	radosgw-admin ratelimit set --ratelimit-scope=user --uid=<uid>
+
+For example:: 
+
+	radosgw-admin ratelimit get --ratelimit-scope=user --uid=johndoe
```
i'm confused about the two separate examples here. the first one says 'ratelimit set' - should we just remove that one?

ofriedma replied:
I've changed the upper example to get

cbodley commented on doc/radosgw/admin.rst:
```diff
@@ -473,6 +473,174 @@ commands. ::
+
+The global rate limit settings can be manipulated with the ``global ratelimit``
+counterparts of the ``ratelimit set``, ``ratelimit enable``, and ``ratelimit disable``
+commands. ``Per user and per bucket ratelimit configuration is overriding the global configuration``::
```
i don't think this `Per user and per bucket ratelimit` sentence should be highlighted

ofriedma replied:
I've removed the highlight

cbodley reviewed (commented):
could you please add a sentence or two to `ceph/PendingReleaseNotes` to announce this new feature?

ofriedma commented:
> could you please add a sentence or two to `ceph/PendingReleaseNotes` to announce this new feature?

Done, Could you please look at this?

cbodley commented:
jenkins test docs

cbodley commented:
> ERROR: sphinx-autodoc-typehints 1.15.1 has requirement Sphinx>=4, but you'll have sphinx 3.5.4 which is incompatible.
> ...
> Extension error (sphinx_autodoc_typehints):
> Handler <function process_docstring at 0x7fe9b4da0af0> for event 'autodoc-process-docstring' threw an exception (exception: source code not available)

cbodley commented:
jenkins test docs

ofriedma commented:
Tested it on Sepia, on Ubutntu 20.04 I can reproduce this error and on 21.04 the build is working fine, Is there a way to change the apt repository or the Ubuntu version we are using?

cbodley commented:
jenkins test docs

cbodley commented:
@dang can you please revisit your review here?
